### ISSUE TITLE
missing a way to serialise numbers to a canonical stringy form

### ISSUE THREAD

**Issue by mwotton** [christian-marie/phone-numbers#1]
is the omission deliberate? useful for json & database storage etc.

**Comment by mwotton** [comment]
oh wait, show. are show and parsePhoneNumber intended to be isomorphisms?

**Comment by christian-marie** [comment]
Sorry, missed this. I am pretty reluctant to make any guarantees about the isomorphisms of libphonenumber. It's hairy.

**Comment by christian-marie** [comment]
I think you want:

``` java
PhoneNumberUtil phoneUtil = PhoneNumberUtil.getInstance();
String formattedNumber = phoneUtil.format(inputNumber, PhoneNumberFormat.E164);
```

http://stackoverflow.com/questions/21374685/convert-to-e164-only-if-possible

Now you see what we're dealing with, it's not too difficult. Is that what you want?

**Comment by mwotton** [comment]
Yes, I think so. We'll want to bounce anything that doesn't make sense back to the user anyway.

**Comment by jberryman** [comment]
If I'm not mistaken, we can get e164 from the current API with just `("+" <>) `, `refCountryCode` and `refNationalNumber`. Does that seem right?

**Comment by jberryman** [comment]
To answer my own question: "sort of...?". libphonenumber is basically doing above:

https://github.com/googlei18n/libphonenumber/blob/master/java/libphonenumber/src/com/google/i18n/phonenumbers/PhoneNumberUtil.java#L1093

But this isn't guaranteed to be a dialable or valid (I think) number. Maybe if the number passes `isValidNumber` then the formatting is "canonical", but I'm not even sure of that. See my answer to the SO question above.

**Comment by christian-marie** [comment]
So did anyone want me to do something to the API?

**Comment by christian-marie** [comment]
For context, I have simply been using this for canonical in-memory representations. Perhaps exposing format and isValidNumber is what is needed? Would this solve anything?

**Comment by jberryman** [comment]
I think `format` for E.164 is implemented as just `"+" <> countryCode <> nationalNumber`, unless I'm misreading the code, so I don't think it buys me anything personally. I'm just using `refCountryCode` and `refNationalNumber` to convert to E.164 .

But certainly exposing `format` operations would probably useful for folks.

**Comment by christian-marie** [comment]
I am ok with with the present state. Thanks for all your input, and especially the PR @jberryman. Closing this.

**Comment by christian-marie** [comment]
For reference, #3

**Comment by christian-marie** [comment]
Perhaps better documentation could make more references as to implementation output guarantees. Patches welcome. I am not sure why this was closed apologies @mwotton.

P.S. Yes the omission was deliberate. If a different representation should be provided then perhaps this should be the implementation? `"+" <> countryCode <> nationalNumber`? I'm not sure if this implementation would be faster or clearer in some way.

edit: I would be reluctant to include extra representations. If this is a good database representation then perhaps the name could express that along with any guarantees expected of the output and maybe it should be a partial encoding with bad numbers explicitly expressed.
